packages/abi-coder/src/fragments/param-type.ts

```diff
@@ -0,0 +1,175 @@
+      });
+    }
+    populate(this, params);
+
+    const match = this.type.match(paramTypeArray);
+    if (match && match[1]) {
+      populate(this, {
+        arrayLength: parseInt(match[2] || '-1', 10),
+        arrayChildren: ParamType.fromObject({
+          type: match[1],
+          components: this.components,
+          typeArguments: this.typeArguments,
+        }),
+        baseType: 'array',
+        type: this.type,
+      });
+    } else {
+      populate(this, {
+        arrayLength: undefined,
+        arrayChildren: undefined,
+        baseType: this.components !== undefined ? 'tuple' : this.type,
+        type: this.type,
+      });
+    }
+
+    this.isParamType = true;
+
+    Object.freeze(this);
+  }
+
+  getSignaturePrefix(): string {
+    if (this.type) {
+      const arrayMatch = arrayRegEx.exec(this.type)?.groups;
+      const structMatch = structRegEx.exec(this.type)?.groups;
+      const enumMatch = enumRegEx.exec(this.type)?.groups;
+
+      // TODO: return 'a' after this PR is merged
+      // https://github.com/FuelLabs/fuels-ts/pull/419
+      if (arrayMatch) return '';
+      if (structMatch) return 's';
+      if (enumMatch) return 'e';
+    }
+
+    return '';
+  }
+
+  getSignatureContent(): string {
+    const type = this.type || '';
+    const arrayMatch = arrayRegEx.exec(type)?.groups;
+    if (arrayMatch) {
+      return `[${arrayMatch.item}; ${arrayMatch.length}]`;
+    }
+
+    if (this.baseType === 'array' && this.arrayChildren) {
+      return `${this.arrayChildren.getSighashSignature()}[${
+        this.arrayLength && this.arrayLength >= 0 ? String(this.arrayLength) : ''
+      }]`;
+    }
+
+    if (this.baseType === 'tuple' && this.components) {
+      const typeArgumentsSignature = this.typeArguments
+        ? `<${this.typeArguments.map((typeArg) => typeArg.getSighashSignature()).join(',')}>`
+        : '';
+      const componentsSignature = `(${this.components
+        .map((comp) => comp.getSighashSignature())
+        .join(',')})`;
+
+      return `${typeArgumentsSignature}${componentsSignature}`;
+    }
+
+    return type;
+  }
+
+  getSighashSignature(): string {
+    const prefix = this.getSignaturePrefix();
+    const content = this.getSignatureContent();
+
+    return `${prefix}${content}`;
+  }
+
+  static fromObject(value: JsonFragmentType | ParamTypeProps): ParamType {
+    if (ParamType.isParamType(value)) {
+      return value;
+    }
+
+    return new ParamType(CONSTRUCTOR_GUARD, {
+      name: value.name,
+      type: value.type,
+      indexed: value.indexed !== undefined ? Boolean(value.indexed) : undefined,
+      components: value.components ? value.components.map(ParamType.fromObject) : undefined,
+      typeArguments: value.typeArguments
+        ? value.typeArguments.map(ParamType.fromObject)
+        : undefined,
+    });
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static isParamType(value?: any): value is ParamType {
```

Comment:
Can we change from `any` to `?: ParamType`? or `value?:  { isParamType: boolean }`?